### ISSUE TITLE
Feature Suggestion - Compatibility

### ISSUE THREAD

**Issue by dazinator** [dotnet/cli#6614]
Idea arisen from: https://github.com/dotnet/cli/issues/6611

My understanding is the CLI is designed to run side by side (multiple versions) and the policy is to always default to the latest version of the CLI unless a version is explicitly provided (global.json sdk value, not sure how it's provided via msbuild as dotnet calls into msbuild, so if sdk version is now an msbuild property I guess it works by msbuild calling dotnet cli, which calls back into msbuild!?)

The problem I have encountered with this, is if you install a newer version of the CLI that has a breaking change, then this can now get selected to build your legacy projects, which are now incompatible with the selected cli, thus builds that were building are not breaking. 

I think it would be better if:

1. A project / solution contained a value indicating the CLI version that it is compatible / produced with.
2. When the cli is executed, pretty much all of the commands require a project / solution as an argument. It can retrieve this value. It can then do a semver comparison to select only the latest "compatible" cli version rather than just the "latest" version.

For example, if I create a project using cli version 2.0.0.
If later I install cli 2.0.1 on a build server - then this will be selected to build the project, because the semantic version number is greater but doesn't indicate that there are any breaking changes from the target project which used cli version 2.0.0.

For this same reason cli version 2.1.0 would later be selected if I happened to install that as well.

However when I install cli version 3.0.0 - this would not be selected for cli commands executed against my cli version 2.0.0 project - as the major version bump indicates a breaking change.

However if you introduce a breaking change with the next version of the CLI, then when people install this next version, the policy may result in the laterversion of the CLI being selected to build legacy projects
Would you consider adding a CLI version msbuild property to projects - to indicate the version of the cli that was used to manage them;

**Comment by dasMulli** [comment]
Related: There is https://github.com/dotnet/core-setup/issues/679 proposing wildcards / min-max ranges in global.json.

The real question is probably why so would want to build using an older CLI (project.json conversion put aside). The only blocker I know of is probably the implicit framework reference, which is being discussed at https://github.com/dotnet/designs/issues/3.

**Comment by dazinator** [comment]
Project.json is the reason for me :-). You couldnt keep compat there, so can you guarantee from now on to always keep new releases compatible? Its a hard guarantee to make so surely better to have atleast a mechanism for control?

**Comment by svick** [comment]
> surely better to have at least a mechanism for control

You already do. It's called global.json.

**Comment by dazinator** [comment]
So is global.json also the mechanism of control for   msbuild based projects moving forward?

As a side note, I checked in a project that has a global.json specifying a particular sdk version. When this was built on a build agent that had a different sdk version installed, the sdk version in global.json was ignored and the latest cli version was used.

**Comment by dsplaisted** [comment]
> As a side note, I checked in a project that has a global.json specifying a particular sdk version. When this was built on a build agent that had a different sdk version installed, the sdk version in global.json was ignored and the latest cli version was used.

Yes, we are planning on fixing that: https://github.com/dotnet/core-setup/issues/2334

**Comment by svick** [comment]
> So is global.json also the mechanism of control for msbuild based projects moving forward?

Yes. That's why it's still in [the documentation](https://docs.microsoft.com/en-us/dotnet/articles/core/tools/global-json), even though project.json (and project.json-relevant properties of global.json) was removed from there.

**Comment by dazinator** [comment]
@svick Ok cool!
I noticed the very first comment on that docs page is someone asking how to specify a range for sdk versions in global.json.

Locking a build to a particular sdk version is needed, and its obviosly understood why that is needed and good that global.json facilitates that. Also good to know that bug will be fixed.

Where this feature suggestion comes in then, is if you dont care exactly which sdk version is used as long as its "a compatible one" for your project / solution. For example why would I want to lock the solution to exactly sdk version 2.0.0 if sdk 2.0.1 comes out next week with a patch that makes the build 10 times faster? Expressing this type of constraint has been the realm of using semver rules, and so thats what I am pushing for here. So basically by default, if no sdk version specified in global.json, then the latest semver compatible sdk version will be used for the target project / solution. If this was the default behaviour, I wouldnt actually need a global.json for most of the solutions I have worked on at all.

**Comment by dsplaisted** [comment]
In the future, we don't intend to break compatibility in the same way we did with the change from project.json to .csproj.  So in general, if you're using .csproj we would recommend letting the latest SDK be used and not using a global.json.  Going forward, the global.json functionality is intended more as a fallback solution in case we break compatibility.  You could also use it if you want to be paranoid about compatibility breaks in the SDK, but in that case you probably don't want to roll forward to patch or minor updates either, as we could accidentally break compat in those as well.

**Comment by dazinator** [comment]
@dsplaisted - thanks for response.

> we don't intend to break compatibility in the same way we did with the change from project.json to .csproj.  So in general, if you're using .csproj we would recommend letting the latest SDK be used and not using a global.json.

and
>  Going forward, the global.json functionality is intended more as a fallback solution in case we break compatibility.

Those two things taken together suggest the recommended path is one that will currently lead to  existing envrionments breaking if a newer, non backwards compatibile cli is installed on the machine. (Which happens currently with VS2015 and VS2017 cli's) without any warning from the installers. At this point, existing builds and processes using the old cli are broken, which obviously cost time and money to investigate and resolve, currently leading to the following solution: developers please add a global.json to all of your solutions now.

>  You could also use it if you want to be paranoid about compatibility breaks in the SDK, 

Well, I don't think it's necessarily paranoid to want to pre-empt and mitigate from compatibility breaks, especially in this case given that they are quite wasteful of peoples time when end up breaking entire environments, and given that releases that have breaking changes are hardly uncommon in the software industry. If the cli team could make a guarantee that there would never ever be another breaking change, then I'd say this might be overly paranoid, but to date I don't think that commitment has been made. So imho this just seems common sense, not paranoi :-)

 > but in that case you probably don't want to roll forward to patch or minor updates either
 
Actually this is what I do want, and what I am trying to suggest this feature for. I want to move forward to patch and minor updates that don't break compatibility. It's only compatibility breaks that cause us down time. Patch and minor updates that keep compatibility are brilliant, but at the moment, if we want to switch to one, we have a lot of churn going through all of our global.json files each time switching the sdk version. And if we don't have a global.json, then we aren't protected from breaking changes in the cli.
